feat: CLDSRV-357 update versioningPreprocessing() helper for null keys

jonathan-gramain opened the pull request:
- Modify the code flow of versioningPreprocessing() to support null keys in addition to the legacy "null versioned keys".
    
  NOTE: The null version compatibility mode is still enforced for now until all pieces are updated to make functional tests pass without the compatibility mode.

- pass deleteNullKey param to backend

  Pass the 'deleteNullKey' param that processVersioningState() may set to the Metadata backend, which tells it to delete the null key as the PUT is executed.

jonathan-gramain committed:
feat: CLDSRV-357 update versioningPreprocessing() helper for null keys

Modify the code flow of versioningPreprocessing() to support null keys
in addition to the legacy "null versioned keys".

NOTE: The null version compatibility mode is still enforced for now
until all pieces are updated to make functional tests pass without the
compatibility mode.

jonathan-gramain committed:
feat: CLDSRV-357 pass deleteNullKey param to backend

Pass the 'deleteNullKey' param that processVersioningState() may set
to the Metadata backend, which tells it to delete the null key as the
PUT is executed.

jonathan-gramain committed:
CLDSRV-357 honor 'delOptions.deleteData' even if master is null

Move check of 'delOptions.deleteData' in prepareNullVersionDeletion()
earlier, so that it is honored even if the master key is a null
version.

This goes with the new possibility to return 'delOptions' without
deleteData in order to delete an existing null key for the master key
(done as part of CLDSRV-353).

tmacro commented on lib/api/apiUtils/object/versioning.js:
```diff
@@ -83,50 +85,54 @@ function _storeNullVersionMD(bucketName, objKey, objMD, options, log, cb) {
+                     { bucketName, objKey, error: err, method: '_deleteNullVersionMD' });
+            return cb(err);
+        }
+        return cb();
```
```suggestion
        }
        return cb(err);
```

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-357 contains:


* `7.70.20`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.20`

* `8.4.20`

* `8.5.21`

* `8.6.9`

* `8.7.12`


Please check the `Fix Version/s` of CLDSRV-357, or the target
branch of this pull request.

jonathan-gramain commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport` with contents from `feature/CLDSRV-357-versioningPreprocessingNullKeySupport`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport origin/development/8.4
 $ git merge origin/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport` with contents from `w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport origin/development/8.5
 $ git merge origin/w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
```

jonathan-gramain committed:
CLDSRV-357 fix behavior change causing issue with mongo backend

When looking for a null version to delete, the code had changed its
behavior: it triggered a delete on the metadata even when the null
version was deleted. While technically not an issue for S3C as
Metadata returns a 200 when attempting to delete a non-existing
version (albeit creating a useless request), the MongoDB backend
returns NoSuchKey error in that case, leading to an issue.

Revert the behavior so to not trigger a delete of the null version
when it does not exist.

jonathan-gramain commented:
/force_reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport` with contents from `feature/CLDSRV-357-versioningPreprocessingNullKeySupport`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport origin/development/8.4
 $ git merge origin/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport` with contents from `w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport`
and `development/8.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport origin/development/8.5
 $ git merge origin/w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `feature/CLDSRV-357-versioningPreprocessingNullKeySupport` into
`development/7.70`
* [w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/feature/CLDSRV-357-versioningPreprocessingNullKeySupport) will be merged into `development/8.4` 
* [w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/feature/CLDSRV-357-versioningPreprocessingNullKeySupport) will be merged into `development/8.5` 
* [w/8.6/feature/CLDSRV-357-versioningPreprocessingNullKeySupport](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/feature/CLDSRV-357-versioningPreprocessingNullKeySupport) will be merged into `development/8.6` 
* [w/8.7/feature/CLDSRV-357-versioningPreprocessingNullKeySupport](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/feature/CLDSRV-357-versioningPreprocessingNullKeySupport) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`



Please check the status of the associated issue CLDSRV-357.

Goodbye jonathan-gramain.
